### ISSUE TITLE
Horovod + TF default configuration can cause CPU threads oversubscription

### ISSUE THREAD

**Issue by nouiz** [horovod/horovod#2295]
Horovod + TF in its default configuration can cause heavy CPU oversubscription. The cause is that by default TF uses for its max intra thread parallelism all the (cores)hyper-threads available on the computer. If there are 8 GPUs, then we end up with 8 processes that each can take all the cores of the machine.

We suggest that Hovorod set TF environments variable TF_NUM_INTRAOP_THREADS to a better default value of nb_cores/nb_gpus. This way by default, it won’t cause CPU oversubscription by default.
There may be cases where the oversubscription is beneficial (like if the different workers have different workloads), so having an Hovorod option that controls that would help to document this and make it easy for users to experiment with this.

@tgaddair

**Comment by nouiz** [comment]
What do you think of this change? Having by default an oversubscription of NGPU*NB_Cores isn't a great.

**Comment by tgaddair** [comment]
Hey @nouiz, thanks for the suggestion!  @romerojosh, do you have any thoughts on this?

**Comment by romerojosh** [comment]
I agree that helping users to avoid oversubscribing CPU resources is a good thing, though I'm not sure setting framework specific environment variables is the right way to do this. A more general approach would be to ensure that `horovodrun` launches processes with CPU bindings limiting each process to an appropriate subset of CPU cores (via the `--map-by` argument in OpenMPI or using `taskset` or `numactl` when launching with gloo). If the processes are launched with these bindings, each process will only "see" their assigned subset of cores and any attempt to use the entire set of CPU cores by the framework will be limited to that set, avoiding any oversubscription.
Binding this way is also a good practice on dual-socket systems as you can also ensure processes are assigned to cores on the socket closest to the assigned GPU.

**Comment by nouiz** [comment]
What you suggest sound a good solution. It is more complicated to implement, but provide gurantee and in addition can speed up some case by using good affinity setting. I have no real experience of that, but it looks good.

**Comment by nouiz** [comment]
This look like still needed. Just a ping to have the bot not close it.

**Comment by tgaddair** [comment]
Thanks @nouiz, added enhancement label to prevent closing.

@shirosankaku @mshiryaev do you have thoughts on this issue?
